contracts/Everlong.sol

```diff
@@ -387,25 +407,108 @@ contract Everlong is IEverlong {
+                asBase,
+                position,
+                0,
+                ""
+            );
             _portfolio.handleClosePosition();
         }
     }
 
     /// @dev Close positions until the targeted amount of output is received.
-    /// @param _targetOutput Minimum amount of proceeds to receive.
+    /// @param _targetOutput Target amount of proceeds to receive.
     /// @return output Total output received from closed positions.
     function _closePositions(
         uint256 _targetOutput
     ) internal returns (uint256 output) {
+        // Round `_targetOutput` up to Hyperdrive's minimum transaction amount.
+        _targetOutput = _targetOutput.max(
+            IHyperdrive(hyperdrive).getPoolConfig().minimumTransactionAmount
+        );
+
+        // Since multiple position's worth of bonds may need to be closed,
+        // iterate through each position starting with the most mature.
+        //
+        // For each position, use the expected output of closing the entire
+        // position to estimate the amount of bonds to sell for a partial closure.
+        IEverlong.Position memory position;
+        uint256 totalPositionValue;
+        uint256 expectedValuePerBond;
+        uint256 bondsNeeded;
         while (!_portfolio.isEmpty() && output < _targetOutput) {
-            output += IHyperdrive(hyperdrive).closeLong(
+            // Retrieve the most mature position.
+            position = _portfolio.head();
+
+            // Estimate the value of the entire position, and use it to derive
+            // the expected output per bond.
+            totalPositionValue = IHyperdrive(hyperdrive).previewCloseLong(
                 asBase,
-                _portfolio.head(),
+                position,
                 ""
             );
-            _portfolio.handleClosePosition();
+            expectedValuePerBond = totalPositionValue.mulDivDown(
+                1e18 - maxCloseLongSlippage,
+                position.bondAmount
+            );
+
+            // Calculate the amount of bonds that must be closed.
+            // Round up to overestimate the amount of bonds.
+            bondsNeeded = (_targetOutput - output).divUp(expectedValuePerBond);
+
+            // Close only part of the position if there are sufficient bonds
+            // to reach the target output without leaving a small amount left.
+            // For this case, the remaining bonds must be worth at least
+            // Hyperdrive's minimum transaction amount or 1% of the target
+            // output, whichever is greater.
+            if (
+                position.bondAmount > bondsNeeded &&
+                (position.bondAmount - bondsNeeded).mulUp(
+                    expectedValuePerBond
+                ) >
+                _targetOutput.mulUp(0.01e18).max(
+                    IHyperdrive(hyperdrive)
+                        .getPoolConfig()
+                        .minimumTransactionAmount
```

Comment:
I'd just store this as an immutable. That will prevent you from needing to call `getPoolConfig` so often. This is also being looped over, so it could be called several times in a single call.

Reply:
So I actually took a quick stab at caching poolconfig yesterday and due to stack-too-deep errors / being used everywhere, it's a bigger change than I initially expected. 

I made a task for it on the task list, cool if I do it later as it's own PR?